### PR TITLE
Add "matrix to the power" operation

### LINKED ISSUE #12: Add "matrix to the power" operation
Add `Matrix::power : (Matrix × u64) -> Matrix` equivalent to $M^n$

### PR DESCRIPTION
**Pull request by mgr0dzicki**
Closes #12.

### CODE REVIEW

**Comment by mgr0dzicki** [comment]
Yo @tudny is that what you want?

**Comment by tudny** [comment]
I think we should optimize `get_shape` by remembering the result somewhere, or check the size during initialization. Just a note, not a comment to your code.

File: src/matrices.rs [hunk 1]
```diff
@@ -232,6 +236,27 @@ impl<T: MatrixNumber> Matrix<T> {
         // We do a little trick and apply `self` twice, as it is more memory efficient
         self.checked_operation_on_two(self, |a, _| operation(a))
     }
+
+    pub fn checked_pow(&self, mut exponent: usize) -> anyhow::Result<Self> {
+        let shape = self.get_shape()?;
+        if shape.0 != shape.1 {
```
Comment (tudny) [inline src/matrices.rs]: I would change it to deconstruct `self.get_shape()` into `(h, w) = self.get_shape()`, as we never need the `shape` itself.

File: src/matrices.rs [hunk 2]
```diff
@@ -530,4 +555,14 @@ mod tests {
...
+
+    #[test]
+    fn test_simple_exponentiation() {
+        let m = im![1, 1; 1, 0];
+        assert_eq!(m.checked_pow(0).unwrap(), im![1, 0; 0, 1]);
+        assert_eq!(m.checked_pow(1).unwrap(), im![1, 1; 1, 0]);
+        assert_eq!(m.checked_pow(2).unwrap(), im![2, 1; 1, 1]);
+        assert_eq!(m.checked_pow(9).unwrap(), im![55, 34; 34, 21]);
+        assert_eq!(m.checked_pow(10).unwrap(), im![89, 55; 55, 34]);
+    }
```
Comment (tudny) [inline src/matrices.rs:579]: I would add a test to check `Rational64` as well

**Review by tudny — CHANGES_REQUESTED** [review]

File: src/matrices.rs (hunk 1 above)
Comment (mgr0dzicki) [reply src/matrices.rs]: fine!

File: src/matrices.rs (hunk 2 above)
Comment (mgr0dzicki) [reply src/matrices.rs:579]: sufficient?

Comment (tudny) [reply src/matrices.rs:579]: Yeah!

**Review by tudny — APPROVED** [review]
lgtm, good work!

**Pull request merged**